### PR TITLE
Separate out potentially useful type lists and variant utility functi…

### PR DESCRIPTION
**Pull request by dholladay00**
…ons to separate header files that could be used downstream w/o needing the variant definition itself.

## PR Summary

@Yurlungur @jhp-lanl @gopsub @jdolence 

This PR includes changes that separates potentially useful type lists and variant utility functions into separate header files that could be used by downstream codes without needing to include the variant type itself. It allows one to start with the final monolithic list of eos types (or any of the intermediate type lists) and build from there using concat, transform etc. Then the final variant's type can be constructed using the variant utility functions. 

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [x] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [x] Update the version in cmake.
- [x] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.

### CODE REVIEW

**Review by Yurlungur — APPROVED** [review]
I approve of exposing this machinery. Just need to make sure tests pass. Probably merits a changelog entry too.

**Review by jhp-lanl — APPROVED** [review]
Looks good to me! And good timing! I think we'll probably be using this here soon (i.e. months) in some of the integration work

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -40,6 +40,8 @@ default:
   id_tokens:
     SITE_ID_TOKEN:
       aud: https://asc-git.lanl.gov
+    CI_JOB_JWT:
```
Comment (dholladay00) [inline .gitlab-ci.yml:43]: @rbberger is this change still needed?

**Comment by dholladay00** [comment]
after comments are addressed by @rbberger we can merge @Yurlungur

File: .gitlab-ci.yml (hunk 1 above)
Comment (rbberger) [reply .gitlab-ci.yml:43]: I have no other pipeline or project that needed that change.

Comment (dholladay00) [reply .gitlab-ci.yml:43]: That was the change  that the CI was asking for on other projects.

Comment (Yurlungur) [reply .gitlab-ci.yml:43]: I'm confused---do we need this change or no?

Comment (rbberger) [reply .gitlab-ci.yml]: No. I don't think so. My theory is that re-git switched over to the new version (the thing that they were warning for many months), which deprecated `CI_JOB_JWT` and that there was some outdated runners still floating around while me and Karen were out. The outdated runner would then complain about not getting the no longer provided token. All it needed was to get that runner updated to accept the new SITE_ID_TOKEN, which seems to have eventually happened.

**Comment by dholladay00** [comment]
This is passing gitlab ci, clicking the button.
